nemo/collections/nlp/models/nlp_model.py

```diff
@@ -312,53 +252,225 @@ def setup(self, stage: str) -> None:
+                for mp_rank in range(1, app_state.model_parallel_size):
+                    os.mkdir(os.path.join(base_path, f'mp_rank_{mp_rank:02d}'))
+                    shutil.move(
+                        os.path.join(base_dir, f'mp_rank_{mp_rank:02d}', 'model_weights.ckpt'),
+                        os.path.join(base_path, f'mp_rank_{mp_rank:02d}'),
+                    )
+                    # clean up leftover directory
+                    shutil.rmtree(os.path.join(base_dir, f'mp_rank_{mp_rank:02d}'))
+
+                # create tar file from base_path
+                self.__make_nemo_file_from_folder(save_path, base_path)
+
+                # clean up base_path
+                shutil.rmtree(base_path)
+
+        elif is_global_rank_zero():
+            return super()._default_save_to(save_path)
+        else:
+            return
+
+    @classmethod
+    def restore_from(
+        cls,
+        restore_path: str,
+        override_config_path: Optional[Union[OmegaConf, str]] = None,
+        map_location: Optional[torch.device] = None,
+        strict: bool = False,
+        return_config: bool = False,
+        trainer: Trainer = None,
+    ):
+        """
+        Restores model instance (weights and configuration) from .nemo file.
+
+        Args:
+            restore_path: path to .nemo file from which model should be instantiated
+            override_config_path: path to a yaml config that will override the internal
+                config file or an OmegaConf / DictConfig object representing the model config.
+            map_location: Optional torch.device() to map the instantiated model to a device.
+                By default (None), it will select a GPU if available, falling back to CPU otherwise.
+            strict: Passed to load_state_dict.
+            return_config: If set to true, will return just the underlying config of the restored
+                model as an OmegaConf DictConfig object without instantiating the model.
+            trainer: Must be passed in to use model parallel .nemo
+
+            Example:
+                ```
+                model = nemo.collections.asr.models.EncDecCTCModel.restore_from('asr.nemo')
+                assert isinstance(model, nemo.collections.asr.models.EncDecCTCModel)
+                ```
+
+        Returns:
+            An instance of type cls or its underlying config (if return_config is set).
+        """
+        if not os.path.exists(restore_path):
+            raise FileNotFoundError(f"Can't find {restore_path}")
+
+        app_state = AppState()
+        app_state.model_restore_path = os.path.abspath(os.path.expanduser(restore_path))
+
+        # detect if we have a model parallel .nemo file
+        with tempfile.TemporaryDirectory() as tmpdir:
+            cwd = os.getcwd()
+            os.chdir(tmpdir)
+            cls.__unpack_nemo_file(path2file=restore_path, out_folder=tmpdir)
+            mp_ranks = glob.glob(os.path.join(tmpdir, 'mp_rank*'))
```

Comment:
Isnt this unpack check super expensive for say non-model parallel - large BERT models? No other choice?

Reply:
Can this be moved below so that its done only if `app_state.model_parallel_size is not None`?

Reply:
This is used to detect model parallel. 

The other option is to have users pass in model parallel size via an argument. I was not going this route because I wanted to make model parallelism automatic in NeMo.

Reply:
Hmm youre right we should automate it. We might need to think of a more efficient way - as in just directly using tarfile to open the file and parse the number of folders or something in the future. The extraction step here will be executed on every rank, therefore all ranks will waste a very long time unpacking large chunk of data just to find number of rank files inside it.

Can be done later though. Leaving the comment open as a todo reminder for future work